osquery/events/darwin/iokit.cpp

```diff
@@ -0,0 +1,235 @@
+/*
+ *  Copyright (c) 2014, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+#include <IOKit/IOMessage.h>
+
+#include "osquery/core/conversions.h"
+#include "osquery/events/darwin/iokit.h"
+#include "osquery/tables/system/darwin/iokit_utils.h"
+
+namespace osquery {
+
+REGISTER(IOKitEventPublisher, "event_publisher", "iokit");
+
+struct DeviceTracker : private boost::noncopyable {
+ public:
+  IOKitEventPublisher* publisher{nullptr};
+  io_object_t notification{0};
+
+  explicit DeviceTracker(IOKitEventPublisher* p) : publisher(p) {}
+};
+
+void IOKitEventPublisher::configure() {}
+
+void IOKitEventPublisher::restart() {
+  if (run_loop_ == nullptr) {
+    // There is no run loop to restart.
+    return;
+  }
+
+  // Remove any existing stream.
+  stop();
+
+  port_ = IONotificationPortCreate(kIOMasterPortDefault);
+  // Get a run loop source from the created IOKit notification port.
+  auto run_loop_source = IONotificationPortGetRunLoopSource(port_);
+  CFRunLoopAddSource(run_loop_, run_loop_source, kCFRunLoopCommonModes);
```

Comment:
Just wondering what the difference between `kCFRunLoopDefaultMode` and `kCFRunLoopCommonModes` is. My google-fu seems weak currently, but it seems `kCFRunLoopDefaultMode` is a subset of `kCFRunLoopCommonModes`.